site/content/blog/identify-ghost-diff-during-kc-reconcile.md

```diff
@@ -0,0 +1,276 @@
+
+```bash
+$ kubectl describe app pkg-demo
+...
+@@ create deployment/simple-app (apps/v1) namespace: default @@
+      0 + apiVersion: apps/v1
+      1 + kind: Deployment
+      2 + metadata:
+      3 +   annotations:
+      4 +     kbld.k14s.io/images: |
+      5 +       - origins:
+      6 +         - preresolved:
+      7 +             url: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+      8 +         url: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+      9 +   labels:
+     10 +     kapp.k14s.io/app: "1649925743708215040"
+     11 +     kapp.k14s.io/association: v1.22a4cbb25c518f776737777e8407b8d9
+     12 +   name: simple-app
+     13 +   namespace: default
+     14 + spec:
+     15 +   progressDeadlineSeconds: 600
+     16 +   replicas: 2
+     17 +   revisionHistoryLimit:
+...
+```
+
+You will see that it is creating a deployment, service and HPA.
+
+Let the reconcilliation loop run once. After the reconcilliation loop is run, we will see that another configmap has been generated. Now, if we will run `app describe` again, we will see the exact diff.
+
+```bash
+$ kubectl get configmaps | grep pkg-demo
+pkg-demo-ctrl                     1      12m
+pkg-demo-ctrl-change-4vpdq        1      2m24s
+pkg-demo-ctrl-change-rnjxf        1      12m
+
+$ kubectl describe app pkg-demo
+...
+Status:
+  Conditions:
+    Status:                         True
+    Type:                           ReconcileSucceeded
+  Consecutive Reconcile Successes:  2
+  Deploy:
+    Exit Code:   0
+    Finished:    true
+    Started At:  2022-04-14T08:52:35Z
+    Stdout:      Target cluster 'https://10.96.0.1:443' (nodes: minikube)
+@@ update deployment/simple-app (apps/v1) namespace: default @@
+  ...
+125,125     progressDeadlineSeconds: 600
+126     -   replicas: 1
+    126 +   replicas: 2
+127,127     revisionHistoryLimit: 10
+128,128     selector:
+Changes
+Namespace  Name        Kind        Conds.  Age  Op      Op st.  Wait to    Rs  Ri
+default    simple-app  Deployment  2/2 t   10m  update  -       reconcile  ok  -
+Op:      0 create, 0 delete, 1 update, 0 noop, 0 exists
+Wait to: 1 reconcile, 0 delete, 0 noop
+...
+```
+
+Here, as we can see, the change in the number of replicas has resulted in the creation of ghost diffs. This is because HPA has reduced the no. of Replicas since there is no load on the server.
+
+This is how a Package consumer can discover the reason for ghost diffs and take appropriate action. In this case, adding a [rebase rule](https://carvel.dev/kapp/docs/latest/hpa-deployment-rebase/#docs) will remove the ghost diffs.
+
+
```

Comment:
would be nice to add note about us wanting to update app changes to include diff information there so that there is no need to add --diff-changes=true

Reply:
Adopted